utils/firestore.js

```diff
@@ -780,32 +780,168 @@ const retrieveSiteNotifications = async (siteId, isParent) => {
+            if (key === 'checkedIn') participantQuery = participantQuery.where('331584571.266600170.135591601', '==', 353358909);
         }
+
+        return participantQuery;
+    }   
+
+    // This executes each query and pushes the data to the fetchedResults array.
+    const executeQuery = async (query) => {
+        const operator = isParent ? 'in' : '==';
         const snapshot = await (queries['allSiteSearch'] === 'true' ? query.get() : query.where('827220437', operator, siteCode).get());
-        if(snapshot.size !== 0){
-            return snapshot.docs.map(document => document.data());
+        if (snapshot.size !== 0) {
+            snapshot.docs.forEach(doc => {
+                fetchedResults.push(doc.data());
+            });
         }
-        else{
-            return [];
+    };
+
+    // Remove duplicate Connect_ID entries from fetchedResults. Since we have to perform multiple get() requests to handle the array-contains queries, we get some duplicate entries in fetchedResults.
+    // Example: search for 'John Smith' will result in two get() requests: one for 'John' and one for 'Smith'. Both requests return the same participant. Remove the duplicate.
+    const removeDuplicateResults = () => {
+        let uniqueParticipants = new Set();
+
+        fetchedResults = fetchedResults.filter(participant => {
+            if (uniqueParticipants.has(participant.Connect_ID)) {
+                return false;
+            } else {
+                uniqueParticipants.add(participant.Connect_ID);
+                return true;
+            }
+        });
+    };
+
+    // Remove results that don't match the query. Ex: if the query is for 'John Smith', the results include 'John Doe', 'John Smith', and 'Jane Smith'. Remove 'John Doe' and 'Jane Smith' from the results.
+    // Why? We run multiple queries due to the array-contains constraints, end up with results that don't match the query when more than one array-contains operation is executed.
+    const removeMismatchedResults = () => {
+        fetchedResults = fetchedResults.filter(participant => {
+            return (!queries.firstName || participant.query.firstName.includes(queries.firstName)) &&
+                (!queries.lastName || participant.query.lastName.includes(queries.lastName)) &&
+                (!queries.email || participant.query.allEmails.includes(queries.email)) &&
+                (!queries.phone || participant.query.allPhoneNo.includes(queries.phone)) &&
+                (!queries.dob || participant['371067537'] === queries.dob) &&
+                (!queries.connectId || participant.Connect_ID === parseInt(queries.connectId)) &&
+                (!queries.token || participant.token === queries.token) &&
+                (!queries.studyId || participant.state.studyId === queries.studyId) &&
+                (!queries.checkedIn || participant['331584571.266600170.135591601'] === 353358909);
```

Comment:
`dob`, `connectId` etc. are common keys used in firstName, lastName and email queries, so they may not need to be checked again after data retrieval.

Reply:
I think you are right about this since we build all of these other params into each query.

Example: `firstName` query also contains any other params as compound .where() clauses (dob, connectId, token, studyId, checkedIn). The same is true for `lastName` query and `phone` or `email` query.
So, this filter probably only needs to check `queries.firstName`, `queries.lastName`, `queries.email`, and `queries.phone`.

I need to update this function in the next dev cycle to remove string search support. This will happen once all existing participants are converted to the query.name as array data structures. I will test with the seemingly unnecessary filters removed when I make these updates in the upcoming dev cycle.